### ISSUE TITLE
Introduce DWARF-enabled Haskell package set

### ISSUE THREAD

**Issue by bgamari** [NixOS/nixpkgs#51987]
## Issue description

While `haskell.lib.enableDWARFDebugging` now exists, in practice it's not very useful since stack unwinding rarely succeeds unless all libraries in the process's image have unwind information. I think what we really want is a whole package set: `haskell.packages.ghc864.withDwarf`.

I thought implementing this would be straightforward; unfortunately it seems I was wrong. My attempt was:
```nix
let
  withDwarf = haskellPackages: haskellPackages.override { 
    overrides = self: super: {
      mkDerivation = xs: pkgs.haskell.lib.enableDWARFDebugging (super.mkDerivation xs);
    };
  };
```
Unfortunately this gives rise to
```
error: anonymous function at /nix/store/4yqnapjyaq2ag9idddvpyvsjy1q0diiy-nixpkgs-18.09.783.299814b385d/nixpkgs/pkgs/development/haskell-modules/generic-builder.nix:12:1 called with unexpected argument 'mkDerivation', at /nix/store/4yqnapjyaq2ag9idddvpyvsjy1q0diiy-nixpkgs-18.09.783.299814b385d/nixpkgs/lib/customisation.nix:69:12
```

**Comment by bgamari** [comment]
Ahh, here we are:

```nix
let
  withDwarf = haskellPackages: haskellPackages.override {
    overrides = self: super: {
      mkDerivation = xs: super.mkDerivation (xs //
        { configureFlags = (xs.configureFlags or "") + "--ghc-options=-g --disable-executable-stripping --disable-library-stripping"; }
      );
    };
  };
```

**Comment by matthewbauer** [comment]
Have you looked at separate-debug-info.sh? 

https://github.com/NixOS/nixpkgs/blob/master/pkgs/build-support/setup-hooks/separate-debug-info.sh

It would put debug info into a separate output. That would avoid the additional package set.

**Comment by nh2** [comment]
> Have you looked at separate-debug-info.sh?

Will that work easily for GHC?

I noticed that on https://ghc.haskell.org/trac/ghc/wiki/DWARF#OpenIssues it says:

> Tibell suggested, the best approach might be to look at options to implement `-gsplit-dwarf`, which would allow us to ship a DWARF package separately.

which sounds to me like it won't work straight away.

**Comment by matthewbauer** [comment]
Yeah it will probably require modifications to some things but IMO a separate output makes more sense in this case than a separate package set.

**Comment by bgamari** [comment]
> Yeah it will probably require modifications to some things but IMO a separate output makes more sense in this case than a separate package set.

That's a fair point.

**Comment by nh2** [comment]
I'm all for what we can get as soon as possible.

If we can get the separate output easily, let's do it. If it's a lot of work and we need to do more compiling until we get there, let's do that instead.

**Comment by nh2** [comment]
> `{ configureFlags = (xs.configureFlags or "") + "--ghc-options=-g --disable-executable-stripping --disable-library-stripping"; }`

I think that needs an extra space, like `+ " --ghc-options=..."`, otherwise it may append directly to the last word in `xs.configureFlags`.
